core/src/main/java/com/linecorp/armeria/internal/common/ConnectionEventState.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2024 LY Corporation
+ *
+ * LY Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common;
+
+import java.net.InetSocketAddress;
+
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.netty.util.AttributeKey;
+
+/**
+ * A key to identify a connection event.
+ */
+public final class ConnectionEventState {
+    /**
+     * The attribute key of the {@link ConnectionEventState}.
+     */
+    public static final AttributeKey<ConnectionEventState> CONNECTION_EVENT_STATE =
+            AttributeKey.valueOf(ConnectionEventState.class, "CONNECTION_EVENT_STATE");
+    /**
+     * The remote address of the connection.
+     */
+    private final InetSocketAddress remoteAddress;
+    /**
+     * The local address of the connection.
+     */
+    private final InetSocketAddress localAddress;
+    /**
+     * The actual protocol of the connection.
+     */
+    @Nullable
+    private SessionProtocol actualProtocol;
+
+    /**
+     * The desired protocol of the connection. This field is only used for client-side connection.
+     */
+    @Nullable
+    private SessionProtocol desiredProtocol;
+
+    /**
+     * The connection keep-alive state.
+     */
+    private KeepAliveState keepAliveState = KeepAliveState.NONE;
+
+    /**
+     * Creates a new instance.
+     */
+    public ConnectionEventState(InetSocketAddress remoteAddress,
```

Comment:
I feel like maintaining the state is doing us more harm than good at this point. What do you think of just removing `ConnectionEventState` and using `ChannelUtil#localAddress`, `ChannelUtil#remoteAddress`, `HttpChannelPool#getProtocolIfHealthy` when notifying events instead?